migrations/versions/2022_041211_5269e8c2c2c9_.py

```diff
@@ -0,0 +1,82 @@
+"""Create partner tables and add the partner_id column to users
+
+Revision ID: 5269e8c2c2c9
+Revises: b500363567e3
+Create Date: 2022-04-12 11:56:30.184292
+
+"""
+import arrow
+from alembic import op
+import sqlalchemy as sa
+import sqlalchemy_utils
+
+
+# revision identifiers, used by Alembic.
+revision = '5269e8c2c2c9'
+down_revision = 'b500363567e3'
+branch_labels = None
+depends_on = None
+
+
+def upgrade():
+    partner_table = op.create_table(
+        "partner",
+        sa.Column("id", sa.Integer(), autoincrement=True, nullable=False),
+        sa.Column("name", sa.String(128), nullable=False),
+        sa.Column("contact_email", sa.String(128), nullable=False, unique=True),
+        sa.Column('created_at', sqlalchemy_utils.types.arrow.ArrowType(), nullable=False, server_default=sa.func.now()),
+        sa.Column('updated_at', sqlalchemy_utils.types.arrow.ArrowType(), nullable=True),
+
+        sa.PrimaryKeyConstraint("id"),
+    )
+
+    partner_api_token_table = op.create_table(
+        "partner_api_token",
+        sa.Column("id", sa.BigInteger(), autoincrement=True, nullable=False),
+        sa.Column("token", sa.String(32), nullable=False, unique=True),
+        sa.Column("partner_id", sa.BigInteger(), nullable=False),
+        sa.Column("expiration_time", sqlalchemy_utils.types.arrow.ArrowType(), nullable=True),
+        sa.Column('created_at', sqlalchemy_utils.types.arrow.ArrowType(), nullable=False),
+        sa.Column('updated_at', sqlalchemy_utils.types.arrow.ArrowType(), nullable=True),
+
+        sa.Index("partner_api_token_token_idx", 'token'),
+        sa.Index("partner_api_token_partner_id_idx", 'partner_id'),
+        sa.ForeignKeyConstraint(['partner_id'], ['partner.id'], ondelete='cascade'),
+        sa.PrimaryKeyConstraint("id"),
+    )
+
+    partner_user = op.create_table(
+        "partner_user",
+        sa.Column("id", sa.BigInteger(), autoincrement=True, nullable=False),
+        sa.Column("user_id", sa.BigInteger(), nullable=False),
+        sa.Column("partner_id", sa.BigInteger(), nullable=False),
+        sa.Column("partner_email", sa.String(255), nullable=True),
+        sa.Column('created_at', sqlalchemy_utils.types.arrow.ArrowType(), nullable=False),
+        sa.Column('updated_at', sqlalchemy_utils.types.arrow.ArrowType(), nullable=True),
+
+        sa.Index("partner_user_user_id_idx", 'user_id'),
+        sa.Index("partner_user_partner_id_idx", 'partner_id'),
+
+        sa.ForeignKeyConstraint(['partner_id'], ['partner.id'], ondelete='cascade'),
+        sa.ForeignKeyConstraint(['user_id'], ['users.id'], ondelete='cascade'),
+        sa.PrimaryKeyConstraint("id"),
+    )
+
+    op.add_column('users', sa.Column('partner_id', sa.BigInteger(), nullable=True, server_default=None))
+    op.add_column('users', sa.Column('partner_user_id', sa.String(128), nullable=True, server_default=None))
+
+    op.execute(
```

Comment:
Self hosting user wouldn't need proton so we can remove it here I think.

Reply:
We will need to make sure we mark somehow the Proton entry so the partner id matches:
https://github.com/cquintana92/app/blob/feature/add-login-with-proton/app/proton/proton_callback_handler.py#L12

Or provide a deterministic method for retrieving that entry in the `proton_callback_handler.py` file. Do we have an implementation like that anywhere in the code?

Reply:
I've added a check that only adds the Proton partner if the `CONNECT_WITH_PROTON` variable is set

Reply:
The migration code is quite special and I suggest not to modify it and leave alembic to generate it. The migration should only focus on the database structure and not add/remove data. 

I think we'll add the partner manually when we deploy the code (this operation is to be done once anyway) so we don't need to do that in the migration.

Reply:
If models is the source of truth we need to add a fixture to the tests since the tests don't run the migrations.

Reply:
In tests, the database is created via https://github.com/simple-login/app/blob/b9d26d46f6d5e2f4ba823d5ecbf93ccfbb17a96f/tests/conftest.py#L37
 which is also based on models.py

Reply:
Yes, I meant that alembic migrations aren't run in tests. DB is derived from models as you say :)

Reply:
The database migration model comes from Django which comes from RoR I believe. It offers a great productivity as we don't have to write sql migration scripts.

As a side note, I wrote a migration library for mysql that's based on SQL files and the migration is actually the source of truth https://pypi.org/project/migration-sql/. Of course no one uses it :D.

Reply:
We should make sure the tests run the migrations just to make sure we don't have anything broken. We could use the models to generate the migrations, but then run the tests with the migrations to make sure we won't break production.

Reply:
That's indeed a good idea, I created this PR for using the migration everywhere and remove the shortcut `Base.metadata.create_all(engine)`

Reply:
:+1: